proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,393 @@
+- Community:
+    - Does the library have an active development community?
+    - Does it have a large userbase?
+- Bindings:
+    - Does the library have a Haskell FFI-capable C API?
+    - How difficult will it be to bind to this library?
+- Auditing:
+    - Has the library been audited for security concerns?
+- Coverage:
+    - What algorithms / primitives does this library implement?
+
+Of these, Botan was selected for several reasons:
+
+- It has a BSD license
+- It has a Haskell FFI-capable C API
+- It offers a broad variety of functionality and algorithms, including post-quantum cryptography
+- It is developed and maintained by an active community
+- It has been [audited](https://botan.randombit.net/releases/audit_1.11.18.pdf) in the past
+- There are unmaintained Haskell bindings targeting Botan 2 / GHC 8 that can be examined / revived
+- Significant implementation work has already been performed in order to validate the feasibility of Botan
+
+The others were not selected for several reasons:
+
+- libgcrypt is C, and so the Haskell FFI can bind directly to it, but it has a GNU LGPL license, and provides less functionality than Botan
+- Crypto++ is C++, and has a Boost license (similar to BSD), but it does not export a C API, and provides less functionality than Botan
+- libsodium / NaCL does not provide the necessary functionality, and there are several Haskell libsodium bindings already (see Appendix C - NaCl / libsodium)
+- NSS has a copyleft license
+
+Bindings to Botan would solve the significant problem of dealing with common cryptography by providing much of the necessary 'cryptographic kitchen sink' via bindings to a suitably performant, suitably licensed, open-source library. The Haskell community would not be required to maintain the Botan cryptography library, only bindings to it, and this can be accomplished readily with some effort which has already begun. Furthermore, the Botan library supports several post-quantum cryptography algorithms, which is highly desirable.
+
+Furthermore, bindings to a dedicated cryptography library may yield a performance benefit, as one would expect a popular C++ crypto library such as Botan to be highly performant. Testing in GHCi with `:set +s` for a requested example of `bcrypt` yielded the following results:
+
+```
+-- Botan
+ghci> bcryptGenerateIO "Fee fi fo fum!" r 16
+"$2a$16$j3Obv.HNT8zQ9OXFXEJeQeq3nT5WmUecPYDF0tc3grNsX9N0uDSCm\NUL"
+(3.73 secs, 996,152 bytes)
+
+-- cryptonite
+ghci> hashPassword 16 ("Fee fi fo fum!" :: ByteString) :: IO ByteString
+"$2b$16$nmHHDe/wnfSl85dIFtqjMuAHxqAmL.JQ6knp8e6kf.X1V9/LCZxUy"
+(8.84 secs, 2,138,266,840 bytes)
+```
+
+It was quick and dirty testing in GHCi so your mileage may vary, but the results indicate that bindings to Botan may be significantly faster and consumes less memory than `crypton/ite`, if other modules / functions are similarly performant. Further implentation, testing, and benchmarking will be necessary to confirm this.
```

Comment:
TBH benchmarks in GHCi are meaningless. The proposal would benefit from benchmarking compiled code via `tasty-bench` or `criterion`.

Reply:
Good idea. I'll put together a proper benchmark of a few things (bcrypt, hashing).